tests/kademlia/testencoding.nim

```diff
@@ -0,0 +1,157 @@
+    check:
+      decoded.key.get() == rec.key.get()
+      decoded.value.get() == rec.value.get()
+      decoded.timeReceived.get() == rec.timeReceived.get()
+
+  test "peer encode/decode":
+    let maddr = MultiAddress.init("/ip4/127.0.0.1/tcp/9000").tryGet()
+    let peer = Peer(id: @[1'u8, 2, 3], addrs: @[maddr], connection: connected)
+    let encoded = peer.encode()
+    let decodedRes = Peer.decode(initProtoBuffer(encoded.buffer))
+    check decodedRes.isSome()
+    let decoded = decodedRes.get()
+    check:
+      decoded.id == peer.id
+      decoded.addrs.len == 1
+      decoded.addrs[0].data == maddr.data
+      decoded.connection == connected
+
+  test "message encode/decode roundtrip":
+    let maddr = MultiAddress.init("/ip4/10.0.0.1/tcp/4001").tryGet()
+    let peer = Peer(id: @[9'u8], addrs: @[maddr], connection: canConnect)
+    let msg = Message(
+      msgType: Opt.some(findNode),
+      key: Opt.some(@[7'u8]),
+      record: Opt.some(
+        Record(
+          key: Opt.some(@[1'u8]), value: Opt.some(@[2'u8]), timeReceived: Opt.some("t")
+        )
+      ),
+      closerPeers: @[peer],
+      providerPeers: @[peer],
+    )
+    let encoded = msg.encode()
+    let decodedRes = Message.decode(encoded.buffer)
+    check decodedRes.isSome()
+    let decoded = decodedRes.get()
+    check:
+      decoded.msgType.get() == findNode
+      decoded.key.get() == @[7'u8]
+      decoded.record.get().key.get() == @[1'u8]
+      decoded.closerPeers.len == 1
+      decoded.providerPeers.len == 1
+
+  test "decode record with missing fields":
+    var pb = initProtoBuffer()
+    # no fields written
+    let res = Record.decode(pb)
+    check res.isSome()
+    let rec = res.get()
+    check rec.key.isNone()
+    check rec.value.isNone()
+    check rec.timeReceived.isNone()
+
+  test "decode peer with missing id (invalid)":
+    var pb = initProtoBuffer()
+    let res = Peer.decode(pb)
+    check res.isNone()
+
+  test "decode peer with invalid connection type":
+    var pb = initProtoBuffer()
+    pb.write(1, @[1'u8, 2, 3]) # id field
+    pb.write(3, uint32(999)) # bogus connection type
+    let res = Peer.decode(pb)
+    check res.isNone()
+
+  test "decode message with invalid msgType":
+    var pb = initProtoBuffer()
+    pb.write(1, uint32(999)) # invalid MessageType
+    let res = Message.decode(pb.buffer)
+    check res.isNone()
+
+  test "decode message with invalid peer in closerPeers":
+    let badPeerBuf = @[0'u8, 1, 2] # junk
+    var pb = initProtoBuffer()
+    pb.write(8, badPeerBuf) # closerPeers field
+    let res = Message.decode(pb.buffer)
+    check res.isNone()
+
+  test "decode message with invalid embedded record":
+    # encode junk data into field 3 (record)
+    var pb = initProtoBuffer()
+    pb.write(1, uint32(putValue)) # valid msgType
+    pb.write(3, @[0x00'u8, 0xFF, 0xAB]) # broken protobuf for record
+    let res = Message.decode(pb.buffer)
+    check res.isSome()
+    check res.unsafeGet().record.isNone()
```

Comment:
I wonder if `Message.decode` should be throwing an error when pb is invalid.

again: `putValue` -> `Message.putValue`